File: languages/python/gates/type-parity/probe.py
#!/usr/bin/env python3
"""gates/type-parity/probe.py — the `python` arm.

Import the staged package BY NAME (through the packaging boundary, not out of the source
tree), call its `<ext>_type_entities()`, and print one JSON line: for each type entity,
the peer's own content hash and a normalisation of its data tree.

`python` holds the entity data as plain dicts already, so this arm's normaliser is the
thinnest of the three. That is exactly why it is written to REFUSE on an unexpected type
rather than to fall through to `str()`: the cheap path here is the one that would quietly
stringify something the other two arms render structurally, and produce a divergence that
is the instrument's rather than the ports'.
"""

from __future__ import annotations

import importlib
import json
import sys

#: DERIVED, NOT TABULATED, and the table it replaced is why.
#:
#: This arm shipped with `SUBJECTS = {"content": (...), "history": (...)}` -- two entries,
#: which reads as configuration. `tools/check-glue.py` counts what it becomes: one entry per
#: extension in one arm per target is **E x T = 1,196 hand-maintained rows** at the full
#: corpus, in files whose whole purpose is to be extension-AGNOSTIC.
#:
#: The mapping was never a decision. It is this target's naming convention applied to a
#: slug, and the convention is already uniform across every cell this target has. So the
#: arm derives it and REFUSES if the derivation misses -- which is strictly better than a
#: table, because a table cannot notice that a cell broke the convention.
MODULE = "entity_{ext}.types"
ENTITIES_FN = "{ext}_type_entities"


def normalise(value):
    """The port's data tree -> canonical JSON. Map keys sorted; see compare.py."""
    if isinstance(value, dict):
        return {str(k): normalise(v) for k, v in sorted(value.items(), key=lambda kv: str(kv[0]))}
    if isinstance(value, (list, tuple)):
        return [normalise(v) for v in value]
    if isinstance(value, (bytes, bytearray)):
        return {"__bytes__": bytes(value).hex()}
    if isinstance(value, bool) or value is None or isinstance(value, (int, float, str)):
        return value
    raise TypeError(
        f"REFUSING: unhandled value type {type(value).__name__!r} in a type entity. "
        "Falling through to str() here would render structurally on the other two arms "
        "and textually on this one, which is a divergence this gate would report as the "
        "PORTS' rather than its own."
    )


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: probe.py <extension-slug>", file=sys.stderr)
        return 2
    ext = sys.argv[1]
    mod_name, fn_name = MODULE.format(ext=ext), ENTITIES_FN.format(ext=ext)

    try:
        mod = importlib.import_module(mod_name)
    except ImportError as exc:
        print(f"REFUSING: cannot import `{mod_name}` for extension `{ext}` ({exc}). The "
              "arm derives the module name from this target's naming convention; a cell "
              "that breaks the convention is a finding, not a missing table entry.",
              file=sys.stderr)
        return 2
    fn = getattr(mod, fn_name, None)
    if fn is None:
        print(f"REFUSING: `{mod_name}` has no `{fn_name}`. Every cell on this target "
              "publishes its materialised type entities under that name; one that does "
              "not is drift the `sdk-surface` gate should have caught.", file=sys.stderr)
        return 2
    entities = fn()

    types = []
    for name, entity in entities:
        types.append({
            "name": name,
            "hash": bytes(entity.hash).hex(),
            "data": normalise(entity.data),
        })

    print(json.dumps({"port": "python", "extension": ext, "types": types}))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: languages/typescript/gates/type-parity/probe.mjs
// gates/type-parity/probe.mjs — the `typescript` arm.
//
// Import the staged package by name (through its `exports` map, not by a path into the
// source tree), call its `<ext>TypeEntities()`, print one JSON line: the peer's own
// content hash per type entity, plus a normalisation of its data tree.
//
// THIS ARM'S NORMALISER DOES THE MOST WORK OF THE THREE, and that is the interesting
// fact about this port. `python` hands back plain dicts and `rust` a `Value` enum;
// `typescript` holds the data as an ECF-TAGGED TREE -- `{kind:"map", pairs:[[k,v]...]}`,
// `{kind:"int", negative, argument}` -- so every value passes through a translation here
// that does not exist on the other two arms.
//
// That asymmetry is why `compare.py` cross-checks `data` against `hash`: the hash is
// computed by the peer's codec and never touches this function, so a bug in this
// translation shows up as a CONTRADICTION rather than as a divergence blamed on a port.

const subject = process.argv[2];

// DERIVED, NOT TABULATED. This arm shipped with a `SUBJECTS` map -- one entry per
// extension, in a file whose whole purpose is to be extension-agnostic, which is E x T =
// 1,196 hand-maintained rows at the full corpus. The mapping was never a decision: it is
// this target's naming convention applied to a slug. Derive it, and REFUSE if the
// derivation misses -- which a table cannot do, because a table cannot notice that a cell
// broke the convention.
if (!subject) {
  console.error("usage: probe.mjs <extension-slug>");
  process.exit(2);
}
const specifier = `@entity-core/extension-${subject}`;
const exportName = `${subject}TypeEntities`;

let mod;
try {
  mod = await import(specifier);
} catch (err) {
  console.error(
    `REFUSING: cannot import \`${specifier}\` for extension \`${subject}\` (${err.message}). ` +
      `The arm derives the specifier from this target's naming convention; a cell that ` +
      `breaks it is a finding, not a missing table entry.`,
  );
  process.exit(2);
}
if (typeof mod[exportName] !== "function") {
  console.error(
    `REFUSING: \`${specifier}\` exports no \`${exportName}\`. Every cell on this target ` +
      `publishes its materialised type entities under that name; one that does not is ` +
      `drift the \`sdk-surface\` gate should have caught.`,
  );
  process.exit(2);
}
const entities = mod[exportName]();

const hex = (bytes) => {
  if (!(bytes instanceof Uint8Array)) {
    // `contentHash`, not `hash`, on this port -- and the chunking-parity sibling paid for
    // learning that. A probe that hexed `undefined` to "" would report identical empty
    // hashes across three arms and read as perfect parity.
    throw new TypeError(`expected a Uint8Array content hash, got ${typeof bytes}`);
  }
  return Buffer.from(bytes).toString("hex");
};

// The ECF value tree -> canonical JSON. REFUSES on an unknown tag rather than falling
// through: an unhandled kind rendered as `undefined` would drop a field silently, and a
// dropped field is exactly the divergence this gate exists to see.
function normalise(v) {
  switch (v?.kind) {
    case "map": {
      const out = {};
      for (const [k, val] of v.pairs) {
        if (k?.kind !== "text") {
          throw new TypeError(`non-text map key of kind ${k?.kind} in a type entity`);
        }
        out[k.value] = normalise(val);
      }
      // Sorted, matching the other arms: the codec re-sorts map keys length-then-lex, so
      // declaration order provably does not reach the bytes. `hash` is the control on
      // whether dropping that distinction was right.
      return Object.fromEntries(Object.keys(out).sort().map((k) => [k, out[k]]));
    }
    case "array":
      return v.items.map(normalise);
    case "text":
      return v.value;
    case "bool":
      return v.value;
    case "null":
      return null;
    case "float":
      return v.value;
    case "int": {
      const n = v.negative ? -1n - BigInt(v.argument) : BigInt(v.argument);
      // Number, to match the other two arms; BigInt does not survive JSON.stringify and
      // the values in a type entity are small. Refuse rather than silently lose one.
      if (n > BigInt(Number.MAX_SAFE_INTEGER) || n < BigInt(Number.MIN_SAFE_INTEGER)) {
        throw new RangeError(`integer ${n} exceeds the safe range; would not survive JSON`);
      }
      return Number(n);
    }
    case "bytes":
      return { __bytes__: Buffer.from(v.value).toString("hex") };
    default:
      throw new TypeError(
        `REFUSING: unhandled ECF kind ${JSON.stringify(v?.kind)} in a type entity. ` +
          `Returning undefined here would drop the field silently.`,
      );
  }
}

const types = entities.map(([name, entity]) => ({
  name,
  hash: hex(entity.contentHash),
  data: normalise(entity.data),
}));

console.log(JSON.stringify({ port: "typescript", extension: subject, types }));
